diff -ur nc.orig/netcat.c nc/netcat.c
--- nc.orig/netcat.c	2009-03-31 13:41:12.000000000 -0400
+++ nc/netcat.c	2009-03-31 14:05:50.000000000 -0400
@@ -316,10 +316,10 @@
 			 */
 			if (uflag) {
 				int rv, plen;
-				char buf[8192];
+				char buf[16384];
 
 				len = sizeof(cliaddr);
-				plen = jflag ? 8192 : 1024;
+				plen = jflag ? 16384 : 2048;
 				rv = recvfrom(s, buf, plen, MSG_PEEK,
 				    (struct sockaddr *)&cliaddr, &len);
 				if (rv < 0)
@@ -690,12 +690,12 @@
 readwrite(int nfd)
 {
 	struct pollfd pfd[2];
-	unsigned char buf[8192];
+	unsigned char buf[16384];
 	int n, wfd = fileno(stdin);
 	int lfd = fileno(stdout);
 	int plen;
 
-	plen = jflag ? 8192 : 1024;
+	plen = jflag ? 16384 : 2048;
 
 	/* Setup Network FD */
 	pfd[0].fd = nfd;
